server/consumer.go

```diff
@@ -5257,23 +5257,30 @@ func gatherSubjectFilters(filter string, filters []string) []string {
 	return filters
 }
 
-// Will check if we are running in the monitor already and if not set the appropriate flag.
-func (o *consumer) checkInMonitor() bool {
+// shouldStartMonitor will return true if we should start a monitor
+// goroutine or will return false if one is already running.
+func (o *consumer) shouldStartMonitor() bool {
 	o.mu.Lock()
 	defer o.mu.Unlock()
 
 	if o.inMonitor {
-		return true
+		return false
 	}
+	o.monitorWg.Add(1)
 	o.inMonitor = true
-	return false
+	return true
 }
 
-// Clear us being in the monitor routine.
+// Clear the monitor running state. The monitor goroutine should
+// call this in a defer to clean up on exit.
 func (o *consumer) clearMonitorRunning() {
 	o.mu.Lock()
 	defer o.mu.Unlock()
-	o.inMonitor = false
+
+	if o.inMonitor {
+		o.monitorWg.Done()
```

Comment:
This is better done when actually exiting the go routine imo.

Reply:
Meaning make it explicit, but this could be ok.

Reply:
I think the main thing is just to make sure `monitorWg` and `inMonitor` are updated at the same time, so they aren't subject to differing goroutine scheduling.